### ISSUE TITLE
Apache/PHP Long processing timeout issue

### ISSUE THREAD

**Issue by searl08** [dvdoug/BoxPacker#187]
Hi,
I have created a test page for **Open Source Box Packer PHP Classes** to check box packing processing times and DVDOUG is included. the link to the page is [https://tekinfowave.com/packer/test_box_pack.php](url). The page uses the current V3 release of DVDOUG and has a number of 'real world' box packing tasks and options. The page includes display of current PHP Version and some key php.ini settings. When **Use DVDOUG Box Packer Library:** is selected, the **Balance Box Weights:** and **Keep Ship Items Flat:** options are shown. When **Balance Box Weights:** is selected and all the default **Ship Items** and **Custom Boxes** are enabled the processing time can go above **2 minutes**, which in a LAMP FastCGI shared server host, will cause a 500 internal server error with a **"End of script output before headers"** error in the Apache error log. 
When I increased the following CGI settings in my local development environment the error disappeared:
FcgidIOTimeout 1200      
FcgidConnectTimeout 1200         
FcgidProcessLifeTime 3600          
FcgidMaxRequestLen 8131072
Let me know if you would like the page code 
Thanks
Skip

**Comment by dvdoug** [comment]
Hi @searl08 

If you wouldn't mind, the code would be really helpful - otherwise I have to do a lot of copy/pasting off that page to make something I can run locally. I assume you've got the items/boxes in an array, that would be the only part I need.

To confirm - it runs well with the weight balancing turned off?

**Comment by dvdoug** [comment]
Oh never mind, I see you already made that available via the debug button. I'll take a look

**Comment by dvdoug** [comment]
Can I please check what PHP version you're using and whether you have xdebug enabled?

**Comment by searl08** [comment]
Hi Doug,

I have attached the file test_box_pack.php and the test_box_pack_arrays.php

[packer.zip](https://github.com/dvdoug/BoxPacker/files/4068022/packer.zip)

Hope this helps.

Skip

**Comment by dvdoug** [comment]
OK so the testcase I added here: https://github.com/dvdoug/BoxPacker/commit/fb1b49292a963fdc8beeb978b888cb6495db74d1 seems to run 0.08 seconds on Travis's copy of PHP 7.2 which is definitely less than 2 minutes.

XDebug massively slows down PHP so that would be my first guess, are you able to confirm whether you have it enabled or not?

**Comment by searl08** [comment]
Hi Doug,

The Page displays the current PHP version at the top, when you first load the page. [https://tekinfowave.com/packer/test_box_pack.php](https://tekinfowave.com/packer/test_box_pack.php) is using PHP 7.2.26. I am not using xdebug .

You should have the previously uploaded pages from packer.zip above. The problem is one of the the default Ship Items "Test Item 3" has a quantity of 150, which is what slows down the processing to the point of failing in the CGI/FastCGI PHP server scenario.

Skip

**Comment by dvdoug** [comment]
Hmm, there's 150 in that testcase though for item 3 and that works fine on Travis's CLI 😢 

Would you mind giving me a full `phpinfo()` dump to look at?

**Comment by searl08** [comment]
Hi Doug,

I notices you have a line for Test Item 3 in the fb1b492 test case:

$packer->addItem(new TestItem('Test Item 3 - 5.5 x 3.75 x .25', 3.75, 5.5, 0.25, 0.09, false), 150);
where the QT is ", 150);" or part of a single line $packer->addItem. Is this how you pass Item  quantities to the packer object?

If so, this was not clear in the documentation test case and you might want to update the [https://www.boxpacker.io/en/stable/getting-started.html ](url), to show this. 

Let me know and I can rewrite the page taking this into consideration.

Thanks Skip

**Comment by dvdoug** [comment]
Ah good spot, that doesn't appear to be documented. It _shouldn't_ make any difference to runtime though vs calling `addItem` repeatedly though, it's just a shortcut for setup: https://github.com/dvdoug/BoxPacker/blob/master/src/Packer.php#L63

**Comment by searl08** [comment]
Hi Doug,

 

I’ll change the code on  https://tekinfowave.com/packer/test_box_pack.php to reflect this and see if it fixes things. I going out now but should have an answer for you tomorrow.

 

Thanks

 

Skip

 <https://tekinfowave.com/> TIW_Logo_Email

**Comment by searl08** [comment]
Hi Doug,

Sorry it took so long for me to get back. I re-coded the page to use the short **$packer->addItem(new TestItem("Test Item 1 - 8" x 4.75" x 1.5"",4.75,8,1.5,0.68,false),25);** with Quantity included format. As you suspected, no change.

To be clear, you must select **Use DVDOUG Box Packer Library:** and **Balance Box Weights:** options, to increase processing time and get the 500 error

Thanks

Skip

**Comment by dvdoug** [comment]
I ran your test page locally on PHP7.2.26 (using the built in web server, not Apache) after making only minimal adjustments to get it running (using a CDN copy of jquery, commenting out the require header/footer) and I get 0m 1sec for the balance box version

```
cd [myworkingdir]
php -S localhost:7000
```
visited http://localhost:7000/test_box_pack.php in the browser.

Can you see if the slowdown reproduces for you outside of Apache?

**Comment by dvdoug** [comment]
For the record although I don't think relevant I also changed `<?` to `<?php` as I have short tags disabled

**Comment by searl08** [comment]
Hi Doug,

 

Can you confirm that, when you run the page on https://tekinfowave.com/packer/test_box_pack.php with Use DVDOUG Box Packer Library: and Balance Box Weights: options selected, you see the increase processing time and intermittently get the 500 error? 

 

I am seeing the following for all Custom Boxes and Ship items selected:

 

1.       Balance Box Weights: not selected  - $packer->setMaxBoxesToBalanceWeight(0); (to disable the weight distribution pass completely.)

a.       BOX PACKER CALC. DURATION: 00 Mins. 02 Secs.

2.       Balance Box Weights: selected  - $packer->setMaxBoxesToBalanceWeight(11); (The custom Box count) or could be not set, to use default 12 boxes or less.

a.       BOX PACKER CALC. DURATION: 01 Mins. 22 Secs.)

 

It seems that with Balance Box Weights: option selected, is what is increasing processing time, to possibly beyond the Apache CGI timeout settings. 

 

Concerning your question about running test page under a different ‘non-apache’ server, I guess the point of all this is, most WordPress servers are shared servers (The environment the for which Shipping method is being is being written)and are a LAMP stack with PHP running as CGI/FastCGI. This said, and the fact that on BlueHost (Provider for https://tekinfowave.com)and probably most shared server environments clients will not be able to change Apache CGI timeout settings.

 

So the question is, a DURATION: 01 Mins. 22 Secs. is a valid processing time for all Custom boxes and Ship items selected and Balance Box Weights: option selected? If so, the 2nd weight distribution pass cannot be used in a shared server environment.

 

Let me know your take on my analysis.

 

Thanks

 

Skip

 <https://tekinfowave.com/> TIW_Logo_Email

**Comment by dvdoug** [comment]
I confirm that I see the timeout on your server with those settings, but the same code runs in 1second on my laptop (albeit running without Apache) so there is a huge disparity there.

I therefore would like to figure out if you see that same result (meaning it's something about Apache specifically), or if it's still slow for you without Apache in which case we can try looking into your PHP settings and figuring out what's different between you and me and the same testcase on Travis which also runs there in under 1 second.

Yes shared hosting will be slower than a dedicated server but if the mobile CPU in a laptop can do it in 1 second, shared hosting should be able to do it seconds, not minutes

**Comment by searl08** [comment]
Hi Doug,

 

Started PHP Development Server -  Steps on Windows:

 

1.       In a command prompt:

a.       cd C:\PHP\PHP7_2_26

b.      php -S localhost:8080 -t C:\inetpub\wwwroot\Tekinfowave_New

                                                               i.      PHP 7.2.26 Development Server started at Thu Jan 23 09:33:10 2020

                                                             ii.      Listening on http://localhost:8080

                                                            iii.      Document root is C:\inetpub\wwwroot\Tekinfowave_New

2.       In Browser

a.       http://localhost:8080/packer/test_box_pack.php

b.      CALC. DURATION: 01 Mins. 51 Secs.

                                                              i.      All Custom Boxes and Ship items selected

                                                            ii.      Use DVDOUG Box Packer Library: selected

                                                          iii.      Balance Box Weights: selected

 

As you can see I am seeing : 01 Mins. 51 Secs. and not 1 second processing time. I still don’t know how you are seeing a 1 sec processing time with the above options set.

 

Let me know.

 

Thanks

 

Skip

 <https://tekinfowave.com/> TIW_Logo_Email

**Comment by dvdoug** [comment]
OK, so it's not Apache specific - that's good, cos it should help debugging it.

Are you able to add a `phpinfo();` call at the top of your test page and attach the results?

**Comment by searl08** [comment]
Hi Doug,

 

Here’s the phpinfo for the Local Dev server running under PHP Development Server.

 

Thanks

 

Skip

 <https://tekinfowave.com/> TIW_Logo_Email

**Comment by dvdoug** [comment]
@searl08 Unfortunately, that doesn't seem to have come through to the Github issue from email - think you'll need to can you add it directly onto the issue page

**Comment by searl08** [comment]
Hi Doug,

Here's the phpInfo from the local PHP Dev Server
[phpinfo.zip](https://github.com/dvdoug/BoxPacker/files/4118486/phpinfo.zip)

Thanks

Skip

**Comment by dvdoug** [comment]
OK, I can reproduce this being slow - but still only under 2.x, not 3.x using that testpage. Any chance you modified the library code at all? (e.g. removing the `int` typehints?). Otherwise I still can't see any other way to reproduce, and I've mimicked my settings to match yours.

The issue is indeed the large number of items which causes the weight distribution to perform a large number of iterations - but because you're using decimal sizes, the numbers are automatically converted to ints by PHP before they enter `Packer` which has the result that in all my tests the weights have been treated as 0 which then means the weight redistribution hasn't needed to do anything. It's only if I multiply everything by 1000 I can reproduce on 3.x

**Comment by searl08** [comment]
Hi Doug,

 

No changes to the library.

 

Would it be better for me to pass weights as interger oz’s instead of decimal lbs? Would this make the weight distribution pass run faster?

 

Thanks

 

Skip

 <https://tekinfowave.com/> TIW_Logo_Email

**Comment by dvdoug** [comment]
You want ints for correctness as that's what the typehints in the library have, and by passing floats your values are getting silently coerced and that's liable to give you bad results. If you stick `declare(strict_types=1)` at the top of your script it will alert you to any mismatches (it's annoying PHP makes the calling side opt in to getting errors rather than than letting the receiving side be in control) 😞 

The library is typehinted for `int` all over because in theory that lets PHP perform optimisations that speed things up, although I'm not sure to what extent in practice the engine does that. PHP7.3 and 7.4 definitely seem quicker on this example than 7.2.

**Comment by dvdoug** [comment]
Side note, I noticed you don't have Opcache turned on which also helps with performance

**Comment by dvdoug** [comment]
v3.5.1 (released 2 mins ago) has some optimisations that might help here compared to v3.4.1, @searl08 can you let me know if you see any difference?

**Comment by searl08** [comment]
Hi Doug,

Just installed v3.5.1 using composer. Is there a way to display the version of Box Packer? If not how can I check to make sure v3.5.1 installed correctly?
I changed the page code to convert all decimal inches/lbs. dimensions to integer mm./grams dimensions with round() for box packer. I've included new page source code in the zip file. 
**[packer_3.5.1.zip](https://github.com/dvdoug/BoxPacker/files/4143155/packer_3.5.1.zip)**
Debug info shows integer mm./grams passed to Box Packer. I convert weights back to lbs/ozs. for BOX PACKER RESULTS.
Bottom line: DVDOUG BOX PACKER CALC. DURATION: **05 Mins. 10 Secs.** with Balance Box Weights: selected  - $packer->setMaxBoxesToBalanceWeight(11); (The custom Box count) running on PHP 7.26 Dev Server. 
This would break in a shared server Apache Fast/CGI environment.
Not sure where to go from here.
Thanks

Skip

**Comment by searl08** [comment]
Hi Doug,

I also added **Pre-Packed Item** checkbox for all Ship Items to allow skipping Box Packing for checked Items.
Thanks
Skip

**Comment by dvdoug** [comment]
`composer.lock` should have a record of precisely what version was installed, the composer CLI usually prints it out too I think.

I get an error in your script

![image](https://user-images.githubusercontent.com/1571110/73600718-322f1980-454c-11ea-830a-c4d7e1be6565.png)

**Comment by dvdoug** [comment]
Should be `$packed_description=$packedItem->getItem()->getDescription();`, but if you're not seeing that error maybe you're not picking up the right code? 🤔

**Comment by searl08** [comment]
Hi Doug,
The composer.lock shows:

```
 "packages": [
        {
            "name": "dvdoug/boxpacker",
            "version": "2.6.4",
            "source": {
                "type": "git",
                "url": "https://github.com/dvdoug/BoxPacker.git",
                "reference": "041630fdc173e28e863beee9eee3c830a1c4055b"
            },
```

Why is that when I followed the composer install instructions?

Skip

**Comment by searl08** [comment]
Hi Doug,

I suspect I'm not using the newest version. My composer environment may be hosed. Is there a composer command to clear composer cache, so the newest version is installed by default, even if you have installed a  previous version?

I downloaded 3.5.2 source code .zip release and overwrote all files/folders in the dvdoug folder. Now:

$packed_description=$packedItem->getItem()->getDescription();

works.

Also **DVDOUG BOX PACKER CALC. DURATION: 00 Mins. 05 Secs.** with **Balance Box Weights: selected - $packer->setMaxBoxesToBalanceWeight(11);** which is significantly better and shouldn't cause a problem on Apache shared server PHP FastCGI environment.

**Observations/Suggestions:**

1.  Add some text to **Getting Started** documentation about all dimensions should be passed as integers converting floats to mm/grams standard using round() function.
2.  Ideally, make library 'units agnostic' and be able to accept float dimensions/weights. Products in Wordpress and I'm sure other E commerce applications have product shipping measurements accept decimal units for Metric or Imperial standards. Float dimensions/weights would be more accurate for packed items and returning live Shipping rates.

Skip

**Comment by searl08** [comment]
Hi Doug,

I have confirmed that the V3.5.2 runs in 2 sec. on my BlueHost shared server PHP FastCGI environment.
[https://tekinfowave.com/packer/test_box_pack.php](url)

**Another Observation/Addition:**

1. Add a Version property that can be returned showing the code version in use due to the composer issues I ran into and the fact that you are still supporting V1 V2 and V3 .

I can tell you obviously care about the Box Packer project and greatly appreciate your help in finding a solution to this issue. I will close the issue.
Thanks again
Skip

**Comment by searl08** [comment]
Hi Doug,

I cleared all my Composer caches and tried a fresh composer install using:
```
composer require dvdoug/boxpacker 
```
V2.6.5 was installed from download.
You should maybe change your composer install command to version specific.

Thanks
Skip

**Comment by dvdoug** [comment]
Composer should pick up the latest version by default if you don't specify one. What does `composer why-not dvdoug/boxpacker:^3.0` say?

**Comment by searl08** [comment]
Hi Doug,

 

C:\inetpub\wwwroot\Tekinfowave_New\packer\box_packer>php -version

PHP 7.2.26 (cli) (built: Dec 17 2019 15:29:44) ( NTS MSVC15 (Visual C++ 2017) x64 )

Copyright (c) 1997-2018 The PHP Group

Zend Engine v3.2.0, Copyright (c) 1998-2018 Zend Technologies

 

 

C:\inetpub\wwwroot\Tekinfowave_New\packer\box_packer>composer why-not dvdoug/boxpacker:^3.0

__root__  -  requires  dvdoug/boxpacker (^2.6)

 

I have a Win 8.1 Apache server with multiple PHP Versions. 5.26, 7.0, 7.2.26. I can switch between them using .bat files which stop and start the requested versions windows service.

 

I have modified the version .bat files to set system path variable to the matching PHP version folder, so the cli php –version command returns correct running PHP version.

 

This may have affected the results I was seeing when trying to get composer to install the latest version of dvdoug.

 

Thanks

 

Skip

 <https://tekinfowave.com/> TIW_Logo_Email

**Comment by dvdoug** [comment]
Yeah, that seems likely. Will assume it's all good then.
